src/object.d

```diff
@@ -159,15 +170,271 @@ auto opEquals(Object lhs, Object rhs)
     return lhs.opEquals(rhs) && rhs.opEquals(lhs);
 }
 
-/************************
-* Returns true if lhs and rhs are equal.
-*/
-auto opEquals(const Object lhs, const Object rhs)
-{
-    // A hack for the moment.
+bool opEquals(T, U)(const T lhs, const U rhs)
+    if (is(T == class) && is(U == class) &&
+        __traits(hasMember, T, "opEquals") &&
+        __traits(hasMember, U, "opEquals") &&
+        !is(typeof(lhs.opEquals(rhs)) == bool) &&
+        !is(typeof(rhs.opEquals(lhs)) == bool))
+{
+    // FIXME. This is a hack.
+    // We shouldn't need to cast away const, and if either lhs' or rhs' opEquals
+    // mutates either object, it's undefined behavior. But before we can remove
+    // this, we need to make it so that TypeInfo and friends have the corect
+    // definitions for opEquals so that they work with the other overload. And
+    // any user code using const objects but which doesn't define opEquals such
+    // that it works with const with the other overload will also break once
+    // this is removed. So, we need to get rid of this, but we need to be
+    // careful about how and when we do it.
     return opEquals(cast()lhs, cast()rhs);
 }
 
+version(unittest) private void _testOpEquals(T, U)(bool expected, T lhs, U rhs, size_t line = __LINE__)
```

Comment:
We have made bad experiences with `version(unittest)` in the past and I think Andrei doesn't really like `version(unittest)` anymore. There was an effort for `StdUnittest` (and `CoreUnittest`), but it seems that didn't work out either, so I suggest to move this into the `unittest` block.

Reply:
When I originally implemented this, I don't think that it worked to have templated functions inside a `unittest` (IIRC, it only worked with one instantiation, making it pointless). Fortunately, that's been fixed. I'll update it accordingly.